### DISCUSSION TITLE
Connection Attempt Timeout

### DISCUSSION THREAD

**Discussion by Exergist** [jchristn/SuperSimpleTcp#42] [Q&A] [answered]
Hello! Is there a way to natively specify a timeout or number of connection attempts for the client before the "No connection could be made because the target machine actively refused it" exception is thrown? 

If not I'm assuming I'd have to resort to something like this:

```csharp
public void Connect()
{
	int timer = 2;
	while (true)
	{
		try
		{
			VA.WriteToLog("attempting connection...", "gray");
			client.Connect();
                        break;
		}
		catch (Exception ex)
		{
			if (ex.Message.Contains("No connection could be made because the target machine actively refused it") == true)
			{
				timer += 2;
				if (timer > connectionTimeout)
				{
					VA.WriteToLog("Error: " + ex.Message, "red");
					break;
				}
			}
			else
			{
				VA.WriteToLog("Error: " + ex.Message, "red");
				break;
			}
		}
	}
}
```
With a `connectionTimeout` of 10 seconds this results in 5 connection attempts.

**Comment by jchristn** [comment, marked as answer]
Hi @Exergist there is a parameter is ```SimpleTcpClient.Settings``` called ```ConnectTimeoutSeconds``` which should accomplish the stated goal of having a connection timeout.

You could wrap ```client.Connect()``` in a ```for``` loop to specify the number of retries, like you have above, but instead of managing a variable for the total number of seconds, you could just manage a variable for the number of retries.

The exception thrown by ```Connect()``` if it's unable to connect is a ```SocketException``` - might be easier to just ```catch (SocketException)``` than to parse the message.  It's unlikely to be thrown for much else.

**Reply by Exergist** [reply]
Thanks for the prompt reply!

So for testing I simplified things as follows:

```csharp
public void TimedConnect()
{
	client.Settings.ConnectTimeoutSeconds = 60;
	try
	{
		client.Connect();
	}
	catch (Exception ex)
	{
		VA.WriteToLog("Error: " + ex.Message, "red");
	}
}
```
Regardless of the `ConnectTimeoutSeconds` I receive the `SocketException` after 2 seconds if the connection is not made.

**Reply by Exergist** [reply]
Shoot, I need to clarify further!

In this scenario the server is not yet listening. Basically I'm planning for the scenario where the client tries to connect with the server not yet started, and I want to give the server a bit of time before the client stops trying to connect automatically.

**Reply by jchristn** [reply]
Hi @Exergist it appears that the ```ConnectTimeoutSeconds``` is not being honored internally when I call ```BeginConnect```.  For now the approach you mentioned earlier should work; I need to rework the internals of this method anyhow to use ```ConnectAsync```, and will add it to my backlog.  Cheers, Joel

**Reply by Exergist** [reply]
Would you like me to open an Issue for this finding?

**Reply by jchristn** [reply]
That would be great, thank you @Exergist !

**Reply by Exergist** [reply]
Opened as Issue #44
